sql/src/test/java/org/apache/druid/sql/calcite/schema/DruidSchemaTest.java

```diff
@@ -237,4 +238,36 @@ public void testGetTableMapFoo2()
     Assert.assertEquals("m1", fields.get(2).getName());
     Assert.assertEquals(SqlTypeName.BIGINT, fields.get(2).getType().getSqlTypeName());
   }
+
+  @Test
+  public void testNullDatasource() throws IOException
+  {
+    Map<DataSegment, SegmentMetadataHolder> segmentsMetadata = schema.getSegmentMetadata();
+    Set<DataSegment> segments = segmentsMetadata.keySet();
+    Assert.assertEquals(segments.size(), 3);
+    final DataSegment segmentToRemove = segments.stream().findFirst().orElse(null);
+    Assert.assertFalse(segmentToRemove == null);
+    schema.removeSegment(segmentToRemove);
+    schema.refreshSegments(segments); // can cause NPE without dataSourceSegments null check in DruidSchema#refreshSegmentsForDataSource
```

Comment:
It's unclear that we remove the only segment of 'foo2' datasource (and not a segment from `foo` datasource) as `getSegmentMetadata` returned Map doesn't guarantee order. Which means it's unclear if we test for `dataSourceSegments == null`

Reply:
yeah, changed to get segment with datasource 'foo2'